src/lib/Inference.hs

```diff
@@ -2156,30 +2200,42 @@ synthTopBlock block = do
   (liftExcept =<<) $ liftDictSynthTraverserM $ traverseGenericE block
 {-# SCC synthTopBlock #-}
 
--- main entrypoint to dictionary synthesizer
-trySynthDict :: (Emits n, Builder m, Fallible1 m, EnvExtender m, EnvReader m)
-             => Type n -> m n (Atom n)
-trySynthDict ty = trySynthDictBlock ty >>= emitBlock
+synthInstanceDef
+  :: (EnvReader m, Fallible1 m) => InstanceDef n -> m n (InstanceDef n)
+synthInstanceDef (InstanceDef className bs params body) = do
+  liftExceptEnvReaderM $ refreshAbs (Abs bs (ListE params `PairE` body))
+    \bs' (ListE params' `PairE` InstanceBody superclasses methods) -> do
+      methods' <- mapM synthTopBlock methods
+      return $ InstanceDef className bs' params' $ InstanceBody superclasses methods'
 
-trySynthDictBlock :: (Fallible1 m, EnvReader m) => Type n -> m n (Block n)
-trySynthDictBlock ty = do
+-- main entrypoint to dictionary synthesizer
+trySynthTerm :: (Fallible1 m, EnvReader m) => Type n -> m n (Atom n)
+trySynthTerm ty = do
   hasInferenceVars ty >>= \case
     True -> throw TypeErr "Can't synthesize a dictionary for a type with inference vars"
     False -> do
-      solutions <- liftSyntherM $ buildBlock do
-                     ty' <- sinkM ty
-                     synthDict ty'
+      synthTy <- typeAsSynthType ty
+      solutions <- liftSyntherM $ synthTerm synthTy
       let solutionsInstance = filter ((== UsedInstance) . snd) solutions
       unless (length solutionsInstance <= 1) $
         throw TypeErr $ "Multiple candidate class dictionaries for: " ++ pprint ty
       case solutions of
         [] -> throw TypeErr $ "Couldn't synthesize a class dictionary for: " ++ pprint ty
-        (d, _):_ -> return d
-{-# SCC trySynthDictBlock #-}
+        (d, _):_ -> cheapNormalize d
```

Comment:
Why do we need to cheap normalize here?

Reply:
Just to reduce code size, if we happen to have e.g. a superclass projection from a concrete instance.

Reply:
Would that really reduce code size? Wouldn't that inline the concrete `Atom` of the superclass?

Reply:
True, there's no guarantee that it'll be smaller, but I think in the common case it will be. The `DictExpr` for the superclass instance will be similar in size to the original instance, but you won't have the projection.